lib/agenda/cancel.js

```diff
@@ -4,20 +4,20 @@ const debug = require('debug')('agenda:cancel');
 /**
  * Cancels any jobs matching the passed MongoDB query, and removes them from the database.
  *  @param {Object} query MongoDB query to use when cancelling
- *  @param {Function} cb callback(error, numRemoved) when cancellation fails or passes
  *  @caller client code, Agenda.purge(), Job.remove()
- *  @returns {undefined}
+ *  @returns {Promise} resolved (numRemoved) when cancellation fails or passes
  */
-module.exports = function(query, cb) {
-  debug('attempting to cancel all Agenda jobs', query);
-  this._collection.deleteMany(query, (error, result) => {
-    if (cb) {
+module.exports = function(query) {
+  return new Promise((resolve, reject) => {
```

Comment:
Why create a new promise when [`deleteMany()` already returns one](http://mongodb.github.io/node-mongodb-native/2.2/api/Collection.html#deleteMany)?

Reply:
`Result.n` isn't a promise though and that's what we're resolving.

Reply:
so you can do `return this._collection.deleteMany(query).then(res => res.result.n)`, no?

Reply:
How's this?

Reply:
Much better, just one detail: you need to `throw err;` instead of `return err;` in the `catch()` handler. The `catch()` handler must either throw an error or return a promise that rejects, otherwise the resulting promise will be resolved. Example:

```javascript
const p = new Promise((resolve, reject) => reject(new Error('woops')));

// The below message will print even though `catch()` handler returned an error
p.
  catch(error => error).
  then(() => console.log('Not an error anymore!'));
```